software/speller/data_collection_platform/dcp/api.py

```diff
@@ -38,38 +39,128 @@ def index():
     # TODO return keyboard layout
     return ' '
 
-
-@bp.route('/char_pressed', methods=['POST', 'GET'])
-def char_pressed():
-    if request.method == 'POST':
-        user_char_pressed = request.form['char_pressed']
-
-
 @bp.route("/openbci/start", methods=['POST'])
 def openbci_start():
     # use a separate process to stream BCI data
     from dcp.bci.stream import stream_bci_api
     from multiprocessing import Process
     with current_app.app_context():
-        p = Process(target=stream_bci_api, args=(shared.bci_processes_states,))
+        # Although the subprocess gets initialized inside this scope, it will persist
+        # bci_processes_states references this subprocess_dict
+        subprocess_dict = shared.manager.dict({
+            'character': None,
+            'phase': None,
+            'frequency': None,
+            'q': deque(),
+            'config_id': None, 
+            'bci_config': None,
+            'state': 'start'
+        })
+        # only pass this subprocess dict to ensure that locks are not too contentious
+        p = Process(target=stream_bci_api, args=(subprocess_dict,))
+        # need to start process before referencing it to obtain the right process_id
         p.start()
-        shared.bci_processes_states[p.pid] = True
+        shared.bci_processes_states[p.pid] = subprocess_dict
+    while subprocess_dict['state'] != 'ready' or subprocess_dict['bci_config'] == None:
+        print('BCI NOT READY YET')
```

Comment:
we should be using a logger instead of print statements, this allows us to preserve logs.

Reply:
**This loop could result in an infinite loop.** The reason is that subprocess_dict["state"] might never be "ready", because `inlet = connect_to_bci()` is a blocking call. Therefore, in the case where the stream couldn't be resolved, it will block until a stream is detected.

Reply:
both are valid points, im open to returning an error response after a certain amount of time. should the subprocess be forced to shutdown in that case?

Reply:
Maybe we can add a timeout here or something and in the case it timed out, we would send the frontned a message to display saying that they need to check their BCI connection.

Reply:
how long should the timeout be? and i assume well force-kill the subprocess?

Reply:
I think we can kill the process and give a timeout of 10 seconds would be good, btw for logging, simply copy paste last year's logging config at the top of `api.py` should be fine.